synapse/storage/types.py

```diff
@@ -12,30 +12,49 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from typing import Any, Iterable, Iterator, List, Optional, Tuple
+from typing import Any, Iterator, List, Mapping, Optional, Sequence, Tuple, Union
 
 from typing_extensions import Protocol
 
 """
 Some very basic protocol definitions for the DB-API2 classes specified in PEP-249
 """
 
+_Parameters = Union[Sequence[Any], Mapping[str, Any]]
+
 
 class Cursor(Protocol):
-    def execute(self, sql: str, parameters: Iterable[Any] = ...) -> Any:
+    def execute(self, sql: str, parameters: _Parameters = ...) -> Any:
         ...
 
-    def executemany(self, sql: str, parameters: Iterable[Iterable[Any]]) -> Any:
+    def executemany(self, sql: str, parameters: Sequence[_Parameters]) -> Any:
         ...
 
-    def fetchall(self) -> List[Tuple]:
+    def fetchone(self) -> Optional[Tuple]:
         ...
 
-    def fetchone(self) -> Tuple:
+    def fetchmany(self, size: Optional[int] = ...) -> List[Tuple]:
+        ...
+
+    def fetchall(self) -> List[Tuple]:
         ...
 
     @property
-    def description(self) -> Any:
+    def description(
+        self,
+    ) -> Optional[
+        Sequence[
+            Tuple[
+                Any,
+                Any,
+                Optional[Any],
+                Optional[Any],
+                Optional[Any],
+                Optional[Any],
+                Optional[Any],
```

Comment:
It would be nice if these had more specific types than `Any` -- it seems the first is a `str` and some of the others are `Optional[int]`?

Reply:
Can you find a source on it? Because I couldn't find a better one than the PEP and some other `dbapi`-esc database files.

Reply:
I'm fairly certain the column names must be strings. The others I was simply basing it off the names, but as we don't use them in synapse it most likely doesn't matter.

Interestingly [sqlite3's `description`](https://docs.python.org/3/library/sqlite3.html#sqlite3.Cursor.description) also has `None` as the second field.

Reply:
...i'll make the second one `Optional` as well then, i'll make the first `str`, and the last 5 `int`, does that sound good?
I'll also add a note that the exact types of those 7 tuple-indexes are relatively undefined, in case we encounter a problem like this one again in the future.